### PR TITLE
[Merged by Bors] - Add check for unused dependencies

### LINKED ISSUE #415: CI should check for unused dependencies
Currently `cargo-udeps` detects the following unused dependencies:
```
`fluvio-cli v0.2.0 (/home/ubuntu/fluvio/src/cli)`
└─── dependencies
     ├─── "async-h1"
     ├─── "http-types"
     └─── "rpassword"
`fluvio-sc v0.2.0 (/home/ubuntu/fluvio/src/sc)`
└─── dependencies
     ├─── "toml"
     └─── "utils"
`fluvio-spu v0.1.0 (/home/ubuntu/fluvio/src/spu)`
└─── dependencies
     ├─── "serde_yaml"
     ├─── "toml"
     └─── "utils"
`fluvio-stream-dispatcher v0.2.0 (/home/ubuntu/fluvio/src/stream-dispatcher)`
└─── dev-dependencies
     └─── "flv-util"
`fluvio-stream-model v0.2.0 (/home/ubuntu/fluvio/src/stream-model)`
└─── dependencies
     └─── "fluvio-types"
`fluvio-system-util v0.1.0 (/home/ubuntu/fluvio/src/utils)`
└─── dependencies
     └─── "regex"
`fluvio-types v0.1.0 (/home/ubuntu/fluvio/src/types)`
└─── dependencies
     └─── "tracing"
`flv-test v1.0.1 (/home/ubuntu/fluvio/tests/runner)`
└─── dependencies
     ├─── "k8-client"
     └─── "k8-metadata-client"
```

### PR DESCRIPTION
**Pull request by tjtelan**
Resolves #415

Runs in separate workflow. Runs on `staging` branch (when Bors performs PR merge to `master`)

* Install nightly in CI (to be used by udeps)
* Add udeps to Makefile
* Remove unused crates and minor bumped affected crates

### CODE REVIEW

File: .github/workflows/ci.yml [hunk 1]
```diff
@@ -211,9 +211,9 @@ jobs:
     strategy:
       matrix:
         os: [ubuntu-latest]
-        rust: [stable]
+        rust: [stable, nightly]
```
Comment (sehz) [inline .github/workflows/ci.yml:214]: should use include rather than put requirement for nightly for all checks

File: .github/workflows/ci.yml [hunk 2]
```diff
@@ -232,20 +232,23 @@ jobs:
         with:
           key: ${{ matrix.rust-target }}-${{ matrix.check }}
       - name: Fmt
-        if: ${{ matrix.check == 'fmt' }}
+        if: ${{ matrix.rust == 'stable' && matrix.check == 'fmt' }}
```
Comment (sehz) [inline .github/workflows/ci.yml:235]: don't need for checking stable if you use `include` in the matrix

**Review by sehz — CHANGES_REQUESTED** [review]
excellent cleanup. minor changed needed

**Review by sehz — APPROVED** [review]

**Comment by sehz** [comment]
building check dependencies seems to take long.  Can you cache them?

**Comment by tjtelan** [comment]
Here's what I learned after digging into this more. I'm not sure how we want to proceed.

We are using `rust-cache` to configure our caching for this job. 

We get a cache-miss at the beginning of most of the jobs

```
Restoring paths:
    /home/runner/.cargo/bin
    /home/runner/.cargo/.crates2.json
    /home/runner/.cargo/.crates.toml
    /home/runner/.cargo/git
    /home/runner/.cargo/registry/cache
    /home/runner/.cargo/registry/index
    ./target
In directory:
    /home/runner/work/fluvio/fluvio
Using keys:
    v0-rust-x86_64-unknown-linux-gnu-unused-dependencies-check-1.56.0-nightly-x86_64-unknown-linux-gnu-2d2bc94c8c3a-1f638b2bd5788a73a101
    v0-rust-x86_64-unknown-linux-gnu-unused-dependencies-check-1.56.0-nightly-x86_64-unknown-linux-gnu-2d2bc94c8c3a
No cache found.
```

and we're caching the result of our build.

```
Saving paths:
    /home/runner/.cargo/bin
    /home/runner/.cargo/.crates2.json
    /home/runner/.cargo/.crates.toml
    /home/runner/.cargo/git
    /home/runner/.cargo/registry/cache
    /home/runner/.cargo/registry/index
    ./target
In directory:
    /home/runner/work/fluvio/fluvio
Using key:
    v0-rust-x86_64-unknown-linux-gnu-unused-dependencies-check-1.56.0-nightly-x86_64-unknown-linux-gnu-2d2bc94c8c3a-1f638b2bd5788a73a101
/usr/bin/tar --posix --use-compress-program zstd -T0 -cf cache.tzst -P -C /home/runner/work/fluvio/fluvio --files-from manifest.txt
Cache Size: ~465 MB (487138769 B)
Cache saved successfully
```

However, in the rust-cache README, they give note about their [cache effectiveness](https://github.com/Swatinem/rust-cache#cache-effectiveness).

```
Usage with Stable Rust is most effective, as a cache is tied to the Rust version. Using it with Nightly Rust is less effective as it will throw away the cache every day.
```

I think this means that we're just going to have a slower first build every day.

---

We can either accept that as `cargo-udeps` requires running in nightly, or one alternative is we might push this test to run only when we're merging to `master`?

**Comment by sehz** [comment]
Do we still need this?

**Comment by tjtelan** [comment]
> Do we still need this?

Maybe, if we're accepting #415 as valid enough to include in CI. Otherwise we should close it.

We left off on the question as to whether this check is worth the minutes is takes to run?

It's worst-case runtime will always be the first builds of the day, because `rust-cache` performs poorly on `nightly` (a current requirement of `check-udeps`)

---

Personally, I think it would be ok to include. Without a cache-hit, it isn't the longest running check. So I don't expect that it will result in longer builds.

If you agree, I can rebase this.

File: .github/workflows/ci_unused_deps.yml [hunk 3]
```diff
@@ -0,0 +1,46 @@
+name: Unused Rust crate check
+
+permissions:
+  contents: read
+
+concurrency:
+  group: ${{ github.head_ref || 'ci_unused_deps' }}
```
Comment (sehz) [inline .github/workflows/ci_unused_deps.yml:7]: this needs to be in terms of `dep-check-${{ github.ref }}` or something like that.

File: .github/workflows/ci_unused_deps.yml [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+  cancel-in-progress: true
+
+on:
+  pull_request:
+    branches: [master]
```
Comment (sehz) [inline .github/workflows/ci_unused_deps.yml:12]: I think this needs to be run only in staging

**Review by sehz — CHANGES_REQUESTED** [review]

File: .github/workflows/stale.yml [hunk 5]
```diff
@@ -6,21 +6,24 @@ on:
 
 jobs:
```
Comment (morenol) [inline .github/workflows/stale.yml:7]: I think that the changes on this file should not be in this PR

File: makefiles/check.mk [hunk 6]
```diff
@@ -16,6 +16,12 @@ check-clippy: install-clippy install_rustup_target
 	cargo +$(RUSTV) check --all --all-features --tests $(VERBOSE_FLAG) $(TARGET_FLAG)
 	cargo +$(RUSTV) clippy --all --all-features --tests $(VERBOSE_FLAG) -- -D warnings -A clippy::upper_case_acronyms $(TARGET_FLAG)
 
+install-udeps:
+	cargo install cargo-udeps --locked
+
+check-udeps: install-udeps
```
Comment (morenol) [inline makefiles/check.mk:22]: there are some conflicts on this file

**Review by morenol — CHANGES_REQUESTED** [review]

**Review by morenol — APPROVED** [review]

**Comment by morenol** [comment]
@tjtelan  this PR has conflicts in Cargo.lock

**Comment by tjtelan** [comment]
> @tjtelan this PR has conflicts in Cargo.lock

conflict is now fixed

**Review by sehz — APPROVED** [review]
LGTM.  Great

**Pull request closed without merging**